File: integrations/langgraph/cortadel_langgraph/tools.py
"""LangChain tools that let an agent read and write Cortadel memory on its own initiative.

These are ``langchain_core.tools.StructuredTool`` instances — the tool primitive every LangGraph
agent already speaks (``create_react_agent(tools=[...])``, ``ToolNode([...])``). Each is built with
both a ``func`` and a ``coroutine``, so the same tool object works in ``.invoke()`` and
``.ainvoke()`` graphs.

They operate through LangGraph's ``BaseStore``, not through a Cortadel client directly. That is
deliberate: passing ``store=None`` makes a tool resolve the ambient store via
``langgraph.config.get_store()`` at call time, so ``create_react_agent(store=CortadelStore(...))``
wires the tools up with no further plumbing — and the same tools keep working against an
``InMemoryStore`` in a unit test.
"""
from __future__ import annotations

import uuid
from typing import Any, Optional, Sequence

from langchain_core.tools import StructuredTool
from langgraph.store.base import BaseStore
from pydantic import BaseModel, Field, create_model

from ._namespace import DEFAULT_NAMESPACE, NamespaceLike, resolve_namespace
from ._runtime import ambient_store

__all__ = [
    "AddMemoriesInput",
    "SearchMemoryInput",
    "create_add_memories_tool",
    "create_memory_tools",
    "create_search_memory_tool",
]

_SEARCH_DESCRIPTION = (
    "Search the user's long-term memory for facts relevant to a question. Call this before "
    "answering anything that depends on who the user is, what they have told you before, their "
    "preferences, decisions, projects, or past events. Returns the most relevant stored facts, "
    "or a note that nothing was found."
)

_ADD_DESCRIPTION = (
    "Save durable facts to the user's long-term memory so they are available in future "
    "conversations. Store standalone, self-contained statements — each one should still make "
    "sense read on its own months from now. Do not store transient chatter, and do not store a "
    "fact you already retrieved from memory."
)


# 10 matches the Cortadel SDK's own ``SearchOptions.top_k`` default. An explicit search tool is
# asked for on purpose, so it can afford a wider net than the automatic injection in
# ``CortadelMemory`` (which defaults to 5 because every turn pays for it).
_DEFAULT_TOP_K = 10
_TOP_K_FIELD = dict(ge=1, le=50, description="Maximum number of memories to return.")


class SearchMemoryInput(BaseModel):
    """Arguments for the ``search_memory`` tool."""

    query: str = Field(
        description="What to look for, phrased as a natural-language question or topic."
    )
    top_k: int = Field(default=_DEFAULT_TOP_K, **_TOP_K_FIELD)


class AddMemoriesInput(BaseModel):
    """Arguments for the ``add_memories`` tool."""

    memories: list[str] = Field(
        description=(
            "One or more standalone facts to remember, each a complete sentence "
            '(e.g. "Prefers async written updates over meetings").'
        )
    )


def create_search_memory_tool(
    *,
    store: Optional[BaseStore] = None,
    namespace: NamespaceLike = DEFAULT_NAMESPACE,
    top_k: int = _DEFAULT_TOP_K,
    name: str = "search_memory",
    description: str = _SEARCH_DESCRIPTION,
) -> StructuredTool:
    """Build the ``search_memory`` tool.

    Args:
        store: Store to read from. ``None`` resolves the ambient store with
            ``langgraph.config.get_store()`` at call time, which requires the graph to have been
            compiled with ``store=``.
        namespace: Namespace to search, possibly templated (``("memories", "{user_id}")``).
            Placeholders are filled from ``config["configurable"]`` on each call.
        top_k: Default number of memories to return when the model does not say. Also the
            default baked into the tool's ``top_k`` argument schema.
        name: Tool name exposed to the model.
        description: Tool description exposed to the model.
    """

    def search_memory(query: str, top_k: int = top_k) -> str:
        resolved = _resolve_store(store)
        # `limit=` is BaseStore's own keyword, not ours — it stays spelled LangGraph's way.
        items = resolved.search(resolve_namespace(namespace), query=query, limit=top_k)
        return _format_memories(items)

    async def asearch_memory(query: str, top_k: int = top_k) -> str:
        resolved = _resolve_store(store)
        items = await resolved.asearch(resolve_namespace(namespace), query=query, limit=top_k)
        return _format_memories(items)

    return StructuredTool.from_function(
        func=search_memory,
        coroutine=asearch_memory,
        name=name,
        description=description,
        # The schema's default is what the model actually gets when it omits `top_k` — pydantic
        # fills it in before the function's own default is ever consulted — so the configured
        # value has to be baked into the schema, not just into the closure. Rename the schema
        # field and this argument together, or a configured top_k is silently ignored.
        args_schema=_search_schema(top_k),
    )


def create_add_memories_tool(
    *,
    store: Optional[BaseStore] = None,
    namespace: NamespaceLike = DEFAULT_NAMESPACE,
    name: str = "add_memories",
    description: str = _ADD_DESCRIPTION,
) -> StructuredTool:
    """Build the ``add_memories`` tool.

    Each fact becomes one ``store.put(namespace, <uuid4>, {"content": fact})``. Against a
    :class:`~cortadel_langgraph.store.CortadelStore` that lands in Cortadel's write pipeline, so
    duplicates and contradictions of existing memories are resolved server-side rather than piling
    up — the tool does not need to check first.

    Args:
        store: Store to write to. ``None`` resolves the ambient store at call time.
        namespace: Namespace to write into, possibly templated.
        name: Tool name exposed to the model.
        description: Tool description exposed to the model.
    """

    def add_memories(memories: list[str]) -> str:
        resolved = _resolve_store(store)
        target = resolve_namespace(namespace)
        stored = 0
        for text in _clean(memories):
            resolved.put(target, str(uuid.uuid4()), {"content": text})
            stored += 1
        return _format_stored(stored)

    async def aadd_memories(memories: list[str]) -> str:
        resolved = _resolve_store(store)
        target = resolve_namespace(namespace)
        stored = 0
        for text in _clean(memories):
            await resolved.aput(target, str(uuid.uuid4()), {"content": text})
            stored += 1
        return _format_stored(stored)

    return StructuredTool.from_function(
        func=add_memories,
        coroutine=aadd_memories,
        name=name,
        description=description,
        args_schema=AddMemoriesInput,
    )


def create_memory_tools(
    *,
    store: Optional[BaseStore] = None,
    namespace: NamespaceLike = DEFAULT_NAMESPACE,
    top_k: int = _DEFAULT_TOP_K,
) -> list[StructuredTool]:
    """Both memory tools, ready to drop into ``create_react_agent(tools=...)``."""
    return [
        create_search_memory_tool(store=store, namespace=namespace, top_k=top_k),
        create_add_memories_tool(store=store, namespace=namespace),
    ]


# ── Helpers ──────────────────────────────────────────────────────────────────────────────────


def _search_schema(default_top_k: int) -> type[SearchMemoryInput]:
    """The search args schema, with ``top_k``'s default matching this tool's configuration."""
    if default_top_k == _DEFAULT_TOP_K:
        return SearchMemoryInput
    return create_model(
        "SearchMemoryInput",
        __base__=SearchMemoryInput,
        top_k=(int, Field(default=default_top_k, **_TOP_K_FIELD)),
    )


def _resolve_store(store: Optional[BaseStore]) -> BaseStore:
    if store is not None:
        return store
    resolved = ambient_store()
    if resolved is None:
        raise RuntimeError(
            "No store is available. Either pass store=CortadelStore(...) when building the tool, "
            "or compile the graph with it: create_react_agent(..., store=CortadelStore(...)) / "
            "StateGraph(...).compile(store=CortadelStore(...))."
        )
    return resolved


def _clean(memories: Sequence[str]) -> list[str]:
    return [text.strip() for text in memories if isinstance(text, str) and text.strip()]


def _format_stored(count: int) -> str:
    if count == 0:
        return "Nothing to store: no non-empty memories were provided."
    return f"Stored {count} " + ("memory." if count == 1 else "memories.")


def _format_memories(items: Sequence[Any]) -> str:
    """Render search results as the plain text that becomes the ``ToolMessage`` content."""
    if not items:
        return "No relevant memories found."
    lines = []
    for index, item in enumerate(items, start=1):
        value = getattr(item, "value", None) or {}
        text = _text_of(value)
        score = getattr(item, "score", None)
        suffix = f" (relevance {score:.3f})" if isinstance(score, (int, float)) else ""
        lines.append(f"{index}. {text}{suffix}")
    return "Relevant memories:\n" + "\n".join(lines)


def _text_of(value: Any) -> str:
    """Pull the human-readable text out of an ``Item.value``, whatever key it landed under."""
    if isinstance(value, str):
        return value
    if isinstance(value, dict):
        for key in ("content", "text", "memory", "data"):
            candidate = value.get(key)
            if isinstance(candidate, str) and candidate.strip():
                return candidate.strip()
        return "; ".join(f"{k}={v}" for k, v in value.items() if k != "id")
    return str(value)

File: integrations/langgraph/src/tools.ts
/**
 * LangChain tools that let an agent read and write Cortadel memory on its own initiative.
 *
 * These are built with `tool()` from `@langchain/core/tools` (`src/tools/index.ts`) — the tool
 * primitive every LangGraph agent already speaks (`createReactAgent({ tools })`, `new ToolNode()`)
 * — with a zod schema, the shape `tool()`'s `ToolWrapperParams.schema` documents.
 *
 * They operate through LangGraph's `BaseStore`, not through a Cortadel client directly. That is
 * deliberate: omitting `store` makes a tool read the store off the `LangGraphRunnableConfig` it is
 * called with, so `createReactAgent({ store: new CortadelStore(...) })` wires the tools up with no
 * further plumbing — and the same tools keep working against an `InMemoryStore` in a unit test.
 *
 * **JS-specific note on where the store comes from.** `tool()` invokes its callback as
 * `func(input, childConfig)` where `childConfig = patchConfig(config, …)`
 * (`@langchain/core/dist/tools/index.js`), so argument two is the runnable config — which inside a
 * graph is a `LangGraphRunnableConfig` carrying `.store`. It is *not* a `ToolRuntime`; that type
 * belongs to `createAgent` in the `langchain` package, and its `store` field is typed as
 * `@langchain/core`'s own key-value `BaseStore<string, unknown>` (the `mset`/`mget` one), which is
 * a different class from LangGraph's. Reading `config.store` — which is exactly what
 * `getStore(config)` does — avoids that trap entirely.
 */
import { tool, type ToolRunnableConfig } from "@langchain/core/tools";
import type { LangGraphRunnableConfig } from "@langchain/langgraph";
import type { BaseStore, SearchItem } from "@langchain/langgraph-checkpoint";
import { z } from "zod";

import { DEFAULT_NAMESPACE, resolveNamespace, type NamespaceLike } from "./namespace.js";
import { ambientStore } from "./runtime.js";

const SEARCH_DESCRIPTION =
  "Search the user's long-term memory for facts relevant to a question. Call this before " +
  "answering anything that depends on who the user is, what they have told you before, their " +
  "preferences, decisions, projects, or past events. Returns the most relevant stored facts, " +
  "or a note that nothing was found.";

const ADD_DESCRIPTION =
  "Save durable facts to the user's long-term memory so they are available in future " +
  "conversations. Store standalone, self-contained statements — each one should still make " +
  "sense read on its own months from now. Do not store transient chatter, and do not store a " +
  "fact you already retrieved from memory.";

/**
 * 10 matches the Cortadel SDK's own `SearchOptions.topK` default. An explicit search tool is asked
 * for on purpose, so it can afford a wider net than the automatic injection in `CortadelMemory`
 * (which defaults to 5 because every turn pays for it).
 */
export const DEFAULT_TOOL_TOP_K = 10;

/** Options shared by both tool factories. */
export interface MemoryToolOptions {
  /**
   * Store to use. Omit to read the store off the `LangGraphRunnableConfig` the tool is called
   * with, which requires the graph to have been compiled with `store`.
   */
  store?: BaseStore;
  /**
   * Namespace to use, possibly templated (`["memories", "{userId}"]`). Placeholders are filled
   * from `config.configurable` on each call.
   */
  namespace?: NamespaceLike;
  /** Tool name exposed to the model. */
  name?: string;
  /** Tool description exposed to the model. */
  description?: string;
}

/** Options for {@link createSearchMemoryTool}. */
export interface SearchMemoryToolOptions extends MemoryToolOptions {
  /**
   * Default number of memories to return when the model does not say. Also the default baked into
   * the tool's `topK` argument schema.
   */
  topK?: number;
}

/**
 * The `search_memory` tool's argument schema, for a given default `topK`.
 *
 * Two things here are load-bearing, and both were found by asserting on the emitted JSON Schema.
 *
 * 1. **The default has to be baked into the schema, not just the closure.** The schema's default
 *    is what the model actually gets when it omits `topK` — zod fills it in during parsing, before
 *    the callback's own fallback is consulted. (The Python leg hit the identical trap with
 *    `StructuredTool` and a pydantic args schema.)
 * 2. **`.default()` alone would make `topK` *required*.** LangChain's `toJsonSchema` routes zod v4
 *    through zod's native `z.toJSONSchema`, which derives the schema from the *output* type — and
 *    a `.default()` field is always present on the output, so it lands in `required` and the model
 *    is obliged to supply it. Adding `.optional()` around the default drops it from `required`
 *    while keeping `default: <n>` advertised, and zod still fills the default in when the key is
 *    absent. This has no Python analogue: pydantic derives the schema from the input side.
 */
function searchMemorySchema(defaultTopK: number) {
  return z.object({
    query: z
      .string()
      .describe("What to look for, phrased as a natural-language question or topic."),
    topK: z
      .number()
      .int()
      .min(1)
      .max(50)
      .default(defaultTopK)
      .optional()
      .describe("Maximum number of memories to return."),
  });
}

/** The `add_memories` tool's argument schema. */
function addMemoriesSchema() {
  return z.object({
    memories: z
      .array(z.string())
      .describe(
        "One or more standalone facts to remember, each a complete sentence " +
          '(e.g. "Prefers async written updates over meetings").',
      ),
  });
}

/** The `search_memory` tool's argument type. */
export type SearchMemoryInput = z.infer<ReturnType<typeof searchMemorySchema>>;

/** The `add_memories` tool's argument type. */
export type AddMemoriesInput = z.infer<ReturnType<typeof addMemoriesSchema>>;

/**
 * Build the `search_memory` tool.
 *
 * @example
 * ```ts
 * const agent = createReactAgent({
 *   llm,
 *   tools: [createSearchMemoryTool({ namespace: ["memories", "e2e-alice"] })],
 *   store: new CortadelStore({ baseUrl, userId: "e2e-alice" }),
 * });
 * ```
 */
export function createSearchMemoryTool(options: SearchMemoryToolOptions = {}) {
  const {
    store,
    namespace = DEFAULT_NAMESPACE,
    topK = DEFAULT_TOOL_TOP_K,
    name = "search_memory",
    description = SEARCH_DESCRIPTION,
  } = options;

  return tool(
    async (input: SearchMemoryInput, config: ToolRunnableConfig): Promise<string> => {
      const resolved = requireStore(store, config);
      const items = await resolved.search(resolveNamespace(namespace, config), {
        query: input.query,
        // `limit` is BaseStore's own option name, not ours — it stays spelled LangGraph's way.
        limit: input.topK ?? topK,
      });
      return formatMemories(items);
    },
    { name, description, schema: searchMemorySchema(topK) },
  );
}

/**
 * Build the `add_memories` tool.
 *
 * Each fact becomes one `store.put(namespace, <uuid>, { content: fact })`. Against a
 * {@link CortadelStore} that lands in Cortadel's write pipeline, so duplicates and contradictions
 * of existing memories are resolved server-side rather than piling up — the tool does not need to
 * check first.
 */
export function createAddMemoriesTool(options: MemoryToolOptions = {}) {
  const {
    store,
    namespace = DEFAULT_NAMESPACE,
    name = "add_memories",
    description = ADD_DESCRIPTION,
  } = options;

  return tool(
    async (input: AddMemoriesInput, config: ToolRunnableConfig): Promise<string> => {
      const resolved = requireStore(store, config);
      const target = resolveNamespace(namespace, config);
      let stored = 0;
      for (const text of clean(input.memories)) {
        await resolved.put(target, crypto.randomUUID(), { content: text });
        stored += 1;
      }
      return formatStored(stored);
    },
    { name, description, schema: addMemoriesSchema() },
  );
}

/**
 * Both memory tools, ready to drop into `createReactAgent({ tools })`.
 *
 * The return type is a mutable tuple on purpose: `as const` would give a `readonly` tuple, which
 * keeps destructuring precise but is *not* assignable to `CreateReactAgentParams["tools"]` — and
 * that is this function's whole reason to exist. A plain tuple gets both.
 */
export function createMemoryTools(
  options: SearchMemoryToolOptions = {},
): [ReturnType<typeof createSearchMemoryTool>, ReturnType<typeof createAddMemoriesTool>] {
  const { topK, ...rest } = options;
  return [createSearchMemoryTool({ ...rest, topK }), createAddMemoriesTool(rest)];
}

// ── Helpers ──────────────────────────────────────────────────────────────────────────────────

function requireStore(store: BaseStore | undefined, config?: ToolRunnableConfig): BaseStore {
  const resolved = store ?? ambientStore(config as LangGraphRunnableConfig | undefined);
  if (resolved === undefined) {
    throw new Error(
      "No store is available. Either pass store: new CortadelStore(...) when building the tool, " +
        "or compile the graph with it: createReactAgent({ ..., store: new CortadelStore(...) }) / " +
        "new StateGraph(...).compile({ store: new CortadelStore(...) }).",
    );
  }
  return resolved;
}

function clean(memories: readonly string[]): string[] {
  return memories
    .filter((text): text is string => typeof text === "string")
    .map((text) => text.trim())
    .filter((text) => text !== "");
}

function formatStored(count: number): string {
  if (count === 0) return "Nothing to store: no non-empty memories were provided.";
  return `Stored ${count} ${count === 1 ? "memory." : "memories."}`;
}

/** Render search results as the plain text that becomes the `ToolMessage` content. */
function formatMemories(items: readonly SearchItem[]): string {
  if (items.length === 0) return "No relevant memories found.";
  const lines = items.map((item, index) => {
    const text = textOfValue(item.value);
    const suffix = typeof item.score === "number" ? ` (relevance ${item.score.toFixed(3)})` : "";
    return `${index + 1}. ${text}${suffix}`;
  });
  return `Relevant memories:\n${lines.join("\n")}`;
}

/** Pull the human-readable text out of an `Item.value`, whatever key it landed under. */
function textOfValue(value: unknown): string {
  if (typeof value === "string") return value;
  if (value !== null && typeof value === "object") {
    const record = value as Record<string, unknown>;
    for (const key of ["content", "text", "memory", "data"]) {
      const candidate = record[key];
      if (typeof candidate === "string" && candidate.trim() !== "") return candidate.trim();
    }
    return Object.entries(record)
      .filter(([key]) => key !== "id")
      .map(([key, entry]) => `${key}=${String(entry)}`)
      .join("; ");
  }
  return String(value);
}
